### PR TITLE
[Model] GPT2ForSequenceClassification model

### PR DESCRIPTION
**Pull request by nie3e**
Support `GPT2ForSequenceClassification` model.
Tested on `nie3e/sentiment-polish-gpt2-small`

Test:
```sh
curl -X 'POST' \
  'http://127.0.0.1:8000/classify' \
  -H 'Content-Type: application/json' \
  -d '{
  "model": "nie3e/sentiment-polish-gpt2-small",
  "input": ["Przestronny hotel , jasny , z dużymi oknami .", "Położony całkiem blisko centrum ."]
}'
```

Result:
```json
{"id":"classify-6619cecdb01a4bf8900df136a9b33b15","object":"list","created":1749994841,"model":"nie3e/sentiment-polish-gpt2-small","data":[{"index":0,"label":"POSITIVE","probs":[6.198883056640625e-06,1.7881393432617188e-07,1.0,7.569789886474609e-06],"num_classes":4},{"index":1,"label":"AMBIGUOUS","probs":[0.00013005733489990234,0.004421234130859375,0.005367279052734375,0.990234375],"num_classes":4}],"usage":{"prompt_tokens":17,"total_tokens":17,"completion_tokens":0,"prompt_tokens_details":null}}
```

### CODE REVIEW

File: vllm/model_executor/models/gpt2.py [hunk 1]
```diff
@@ -318,6 +321,88 @@ def load_weights(self, weights: Iterable[tuple[str,
...
+        self.gpt2.load_weights(weight_filter())
+
+        params_dict = dict(self.named_parameters())
+
+        for name, loaded_weight in self_weights:
+            if name.startswith("score"):
+                param = params_dict[name]
+                weight_loader = getattr(param, "weight_loader",
+                                        default_weight_loader)
+                weight_loader(param, loaded_weight)
```
Comment (Isotr0py) [inline vllm/model_executor/models/gpt2.py]: I think we can use `AutoWeightsLoader` here.

File: vllm/model_executor/models/registry.py [hunk 2]
```diff
@@ -172,7 +172,7 @@
                                             "RobertaForSequenceClassification"),
     "ModernBertForSequenceClassification": ("modernbert",
                                             "ModernBertForSequenceClassification"),
-    "Qwen3ForSequenceClassification": ("qwen3", "Qwen3ForSequenceClassification"), # noqa: E501
```
Comment (Isotr0py) [inline vllm/model_executor/models/registry.py]: We shouldn't remove the Qwen3 embedding model registry.

File: vllm/model_executor/models/gpt2.py [hunk 3]
```diff
@@ -318,6 +321,88 @@ def load_weights(self, weights: Iterable[tuple[str,
...
+    """
+
+    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
+        super().__init__()
+        config = vllm_config.model_config.hf_config
+
+        self.gpt2 = GPT2Model(vllm_config=vllm_config,
+                              prefix=maybe_prefix(prefix, "gpt2"))
+        self.score = nn.Linear(config.n_embd, config.num_labels, bias=False)
+        self.activation = nn.Softmax(dim=-1)
```
Comment (Isotr0py) [inline vllm/model_executor/models/gpt2.py]:
```suggestion
        pooler_config = vllm_config.model_config.pooler_config
        self._pooler = Pooler.from_config_with_defaults(
            pooler_config,
            pooling_type=PoolingType.LAST,
            normalize=False,
            softmax=True)
```
Please use `Pooler` for embedding models.

File: vllm/model_executor/models/gpt2.py [hunk 4]
```diff
@@ -318,6 +321,88 @@ def load_weights(self, weights: Iterable[tuple[str,
...
+            final_shape_tensor = logits[pooled_data_i.shape[0] - 1, :]
+
+            pooled_data_lst.append(final_shape_tensor)
+            offset += prompt_len
+
+        pooled_output = torch.stack(pooled_data_lst)
+
+        scores = self.activation(pooled_output)
+        pooled_outputs = [PoolingSequenceGroupOutput(data) for data in scores]
+        return PoolerOutput(outputs=pooled_outputs)
```
Comment (Isotr0py) [inline vllm/model_executor/models/gpt2.py]: Ditto.

File: vllm/model_executor/models/registry.py (hunk 2 above)
Comment (nie3e) [reply vllm/model_executor/models/registry.py]: Oh course, I don't know why it happened. It will be fixed.

File: vllm/model_executor/models/gpt2.py (hunk 3 above)
Comment (nie3e) [reply vllm/model_executor/models/gpt2.py]: This one might be tricky.

In this case I can see that new `self._pooler` is calling
`SimplePooler`'s forward function that looked like this:
```py
    def forward(
        self,
        hidden_states: torch.Tensor,
        pooling_metadata: PoolingMetadata,
    ) -> PoolerOutput:
        pooled_data = self.extract_states(hidden_states, pooling_metadata)
        pooled_data = self.head(pooled_data, pooling_metadata)
        pooled_outputs = [self.build_output(data) for data in pooled_data]
        return PoolerOutput(outputs=pooled_outputs)
```

I can confirm that `self.extract_states` function does exacly what my
```py
        for prompt_len in prompt_lens:
            pooled_data_i = hidden_states[offset:offset + prompt_len]
```
but after that
```py
pooled_data = self.head(pooled_data, pooling_metadata)
```
is kinda destroying the result.
If we initiate `self._pooler` with `softmax=False` then we will get the same as
```py
logits = self.score(pooled_data_i)
```
But of the `PoolerOutput` type instead of `torch.Tensor`. So we will have to get its outputs data and again use `self.activation` layer.

I mean - this function is called `pooler` but it is not exacly what pooler does (correct me if I am wrong). Should we use `LastPool` object instead of my 
```py
        for prompt_len in prompt_lens:
            pooled_data_i = hidden_states[offset:offset + prompt_len]
```
?
I think I need some assistance in resolving this.

Comment (nie3e) [reply vllm/model_executor/models/gpt2.py]: What do you think about this solution?
```py
    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
        super().__init__()
        config = vllm_config.model_config.hf_config
        self.gpt2 = GPT2Model(vllm_config=vllm_config,
                              prefix=maybe_prefix(prefix, "gpt2"))
        self.score = nn.Linear(config.n_embd, config.num_labels, bias=False)
        self.activation = nn.Softmax(dim=-1)
        self._last_pool = LastPool(normalize=False, softmax=False)
```
```py
    def pooler(
            self,
            hidden_states: torch.Tensor,
            pooling_metadata: PoolingMetadata,
    ) -> Optional[PoolerOutput]:
        pooled_data = self._last_pool.extract_states(hidden_states,
                                                     pooling_metadata)
        logits = self.score(pooled_data)
        scores = self.activation(logits)
        pooled_outputs = [PoolingSequenceGroupOutput(data) for data in scores]
        return PoolerOutput(outputs=pooled_outputs)
```

Comment (Isotr0py) [reply vllm/model_executor/models/gpt2.py]: You should put `self.score` in forwarding instead of `pooler`:
```python3
    def forward(
        self,
        input_ids: torch.Tensor,
        positions: torch.Tensor,
        intermediate_tensors: Optional[IntermediateTensors] = None,
        inputs_embeds: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
        hidden_states = self.gpt2(input_ids=input_ids,
                           position_ids=positions,
                           inputs_embeds=inputs_embeds,
                           intermediate_tensors=intermediate_tensors)
        logits = self.score(hidden_states)
        return logits
```
This should work with my suggested changes:
```python3
    def pooler(
        self,
        hidden_states: torch.Tensor,
        pooling_metadata: PoolingMetadata,
    ) -> Optional[PoolerOutput]:
        return self._pooler(hidden_states, pooling_metadata)

    def forward(
        self,
        input_ids: torch.Tensor,
        positions: torch.Tensor,
        intermediate_tensors: Optional[IntermediateTensors] = None,
        inputs_embeds: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
        hidden_states = self.gpt2(input_ids=input_ids,
                           position_ids=positions,
                           inputs_embeds=inputs_embeds,
                           intermediate_tensors=intermediate_tensors)
        logits = self.score(hidden_states)
        return logits
```

File: vllm/model_executor/models/gpt2.py [hunk 5]
```diff
@@ -318,6 +321,88 @@ def load_weights(self, weights: Iterable[tuple[str,
...
+        model: An instance of GPT2Model used for forward operations.
+        score: A layer for calculating logits.
+        activation: Activation function.
+    """
+
+    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
+        super().__init__()
+        config = vllm_config.model_config.hf_config
+
+        self.gpt2 = GPT2Model(vllm_config=vllm_config,
```
Comment (Isotr0py) [inline vllm/model_executor/models/gpt2.py]:
```suggestion
        self.transformer = GPT2Model(vllm_config=vllm_config,
```

**Comment by nie3e** [comment]
@Isotr0py You are right! Thanks to your help the code looks much simpler!

**Review by Isotr0py — APPROVED** [review]
LGTM! Thanks for your patience!

**Comment by nie3e** [comment]
@Isotr0py one thing before I fix tests. I am not sure if this model is correctly put into `_CROSS_ENCODER_MODELS`. What do you think?

Edit: `_TEXT_GENERATION_MODELS` might be good place.

**Comment by Isotr0py** [comment]
You can put it in `_EMBEDDING_MODELS`.

File: tests/models/registry.py [hunk 6]
```diff
@@ -267,6 +267,7 @@ def check_available_online(
     # [Text-only]
     "BertModel": _HfExamplesInfo("BAAI/bge-base-en-v1.5", v0_only=True),
     "Gemma2Model": _HfExamplesInfo("BAAI/bge-multilingual-gemma2", v0_only=True),  # noqa: E501
+    "GPT2ForSequenceClassification": _HfExamplesInfo("nie3e/sentiment-polish-gpt2-small", v0_only=True),  # noqa: E501
```
Comment (Isotr0py) [inline tests/models/registry.py]:
```suggestion
    "GPT2ForSequenceClassification": _HfExamplesInfo("nie3e/sentiment-polish-gpt2-small"),  # noqa: E501
```
I think we have supported decoder-only embedding models in v1.

Comment (aarnphm) [reply tests/models/registry.py]: recently land

**Comment by nie3e** [comment]
@Isotr0py Can we somehow rerun v1-test?

```
Timed Out
Waited 1m 30s
·
Ran in 3h 0m
```

**Comment by Isotr0py** [comment]
#19872 should fix the CI deadlock, you can merge from main to get the fix.

**Comment by DarkLight1337** [comment]
Can you open another PR to update the Supported Models page?
